Add more Sampler implementations for `LoggingClient/Service`

geminiKim opened the pull request:
**Modifications:**
* Moved Sampler package to `com.linecorp.armeria.common.util`
* Forked the [`RateLimitingSampler`](https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/sampler/RateLimitingSampler.java#L45) from [brave](https://github.com/openzipkin/brave/tree/master/brave/src/main/java/brave/sampler)

**Result:**
* Fix #1960

**Questions:**
I couldn't add [`BoundarySampler`](https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/sampler/BoundarySampler.java#L57), because it needs to `traceId` for `isSampled(long traceId)`.

I think if we want to add `BoundarySampler`, it should modify [`isSampled()`](https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/internal/logging/Sampler.java#L73) in `Sampler` first and it should also be modified where use `Sampler` classes. ([LoggingService](https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/logging/LoggingService.java#L150) ...)


And additional question: Can we get `traceId` from `ServiceRequestContext` of `serve` method in [`LoggingService`](https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/logging/LoggingService.java#L150)?


If it doesn't make sense, you can close PR. But I would appreciate if you answered the question. :D
Thank you and feels sorry for incomplete PR. 🙏

trustin commented:
Re: other sampler implementations:

Could we parameterize `Sampler` and make `isSampled` accept it? i.e.

```java
public interface Sampler<T> {
    static Sampler<Object> random(double rate);
    static Sampler<Object> rateLimited(int samplesPerSecond);
    static <U> Sampler<U> parameterized(Function<? super U, ? extends Sampler<? super U>> function);
    static <U> Sampler<U> boundary(double rate, Function<? super U, Long> hasher);

    boolean isSampled(T obj);
}
```

trustin commented:
@adriancole I'm not sure I understood what boundary sampler is for. Would you mind explain it why it is useful over other samplers?

adriancole commented:
I wouldn't use boundary sampler here. it requires a uniformly random input and acts as a bitmask. Honestly I think twitter over-engineered this as the idea was that you would save resources vs doing other things. But big picture I doubt any single process will get enough traffic to want to do sampling like that.. and if they could not prefer a surge-protected approach (rate-limited) anyway.

adriancole commented:
to answer your question, the only benefit boundry sampler has is overhead. it is stateless outside the rate, and the sampling decision is a bitmask so lightning fast.

however, trace IDs aren't perfect and most people get surprised as they expect that mask to result in fidelity between the target and actual rates. when trace IDs aren't exactly random you can get more or less traces passing than expected in other words. It works out pretty fine in high traffic, but surprises most, which is why we don't use it.

trustin commented:
Thanks a lot for the detailed explanation, @adriancole! Let's skip `BoundarySampler`, then.

geminiKim commented:
Thank you for review, I updated codes.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -80,7 +56,57 @@ public boolean isSampled() {
      * @param rate minimum sample rate is 0.01, or 1% of traces
      */
-    public static Sampler create(float rate) {
+    static Sampler<Object> random(double rate) {
```
I realized that this could be more useful to users if we did:

```java
static <T> Sampler<T> random(double rate) {
    @SuppressWarnings(...)
    final Sampler<T> cast = (Sampler<T>) CountingSampler.create(rate);
    return cast;
}
```

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -80,7 +56,57 @@ public boolean isSampled() {
+     *
+     * @param samplesPerSecond minimum rate-limited is 0 and the max is 2,147,483,647 (max int)
+     */
+    static Sampler<Object> rateLimited(int samplesPerSecond) {
```
I realized that this could be more useful to users if we did:

```java
static <T> Sampler<T> rateLimited(double rate) {
    @SuppressWarnings(...)
    final Sampler<T> cast = (Sampler<T>) RateLimitingSampler.create(rate);
    return cast;
}
```

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -80,7 +56,57 @@ public boolean isSampled() {
+    /**
+     * Returns a sampler that always will be sampled.
+     */
+    static Sampler always() {
```
`<T> Sampler<T> always()`

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -80,7 +56,57 @@ public boolean isSampled() {
+     * Returns a sampler that always will be sampled.
+     */
+    static Sampler always() {
+        return new Sampler() {
```
How about making this a singleton in a utility class? e.g.

```java
public interface Sampler<T> ... {
    static <T> Sampler<T> always() {
        return (Sampler<T>) Samplers.ALWAYS;
    }
    static <T> Sampler<T> never() {
        return (Sampler<T>) Samplers.NEVER;
    }
}

final class Samplers {
    static final Sampler ALWAYS = ...;
    static final Sampler NEVER = ...;
}
```

trustin commented on core/src/main/java/com/linecorp/armeria/internal/logging/CountingSampler.java:
```diff
@@ -46,20 +48,20 @@
  * <p>Forked from brave-core 5.6.3 at d4cbd86e1df75687339da6ec2964d42ab3a8cf14
  */
-final class CountingSampler extends Sampler {
+public final class CountingSampler implements Sampler {
```
Could become package-private again because a user can create it via `Sampler.random()`.

trustin commented on core/src/main/java/com/linecorp/armeria/internal/logging/CountingSampler.java:
```diff
@@ -46,20 +48,20 @@
      *             or 1% of traces
      */
-    public static Sampler create(final float rate) {
+    public static Sampler create(final double rate) {
```
`public` could be removed.

trustin commented on core/src/main/java/com/linecorp/armeria/internal/logging/RateLimitingSampler.java:
```diff
@@ -0,0 +1,169 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
```
Could you insert the license header of the OpenZipkin project here?

```java
/*
 * Copyright 2019 LINE Corporation
 * ...
 */
/*
 * Zipkin license header here
 * ...
 */
```

trustin commented on core/src/main/java/com/linecorp/armeria/internal/logging/RateLimitingSampler.java:
```diff
@@ -0,0 +1,169 @@
+ * <p>Forked from brave-core 5.6.9 at b8c00c594cbf75a33788d3dc990f94b9c6f41c01
+ */
+public final class RateLimitingSampler implements Sampler {
+    public static Sampler create(int tracesPerSecond) {
```
public final class -> final class
public static -> static

trustin commented:
@geminiKim Sorry for a late review. We were on a business trip for the last 7 days. :sweat_smile:

geminiKim commented:
Thank you for specific advice and direction, I updated codes.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -80,7 +53,45 @@ public boolean isSampled() {
      *
```
Can we remove the paragraph above, because we do not provide `BoundarySampler`?

geminiKim replied:
Thank you for review. fiexed in 6459312.

trustin replied:
Could you also remove the paragraph: `The sample returned is good for ....`.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -80,7 +53,45 @@ public boolean isSampled() {
+    /**
+     * Returns a sampler, given a rate-limited on a per-second interval.
+     *
+     * @param samplesPerSecond minimum rate-limited is 0 and the max is 2,147,483,647 (max int)
```
How about:

    @param samplesPerSecond an integer between {@code 0} and {@value Integer#MAX_VALUE}

geminiKim replied:
Thank you for review. fiexed in 6459312.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -80,7 +53,45 @@ public boolean isSampled() {
+    }
+
+    /**
+     *  Returns true if a request should be recorded.
```
- Double spaces between `*` and `Returns`.
- `{@code true}`

geminiKim replied:
Thank you for review. fiexed in 6459312.

minwoox commented on core/src/main/java/com/linecorp/armeria/client/logging/LoggingClient.java:
```diff
@@ -152,7 +153,7 @@ public LoggingClient(Client<I, O> delegate, LogLevel level) {
     @Override
     public O execute(ClientRequestContext ctx, I req) throws Exception {
-        if (sampler.isSampled()) {
+        if (sampler.isSampled(ctx.log())) {
```
Is there any chance that a customized sampler uses the information from the `RequestLog` such as `ResponseHeaders`, etc.?
Then we should check this using listener.

trustin replied:
Wouldn't it increase the overhead because we are gonna have to add the listeners even if the request will not be logged actually?

Actually, this makes me think using `RequestLog` as an input for a `Sampler` doesn't make much sense. Should we just pass `RequestContext`?

minwoox replied:
That's a good idea. Yeah, I agree that it's better to hand in `RequestContext` instead of `RequestLog`. 😄

trustin replied:
OK, then. @geminiKim, could you update the type parameters of `Sampler` from `RequestLog` to `ClientRequestContext` (for `LoggingClient`) and `ServiceRequestContext` (for `LoggingService`) ?

geminiKim replied:
Thank you for sharing your detailed discussion. Updated in 7977716.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -43,35 +43,8 @@
  * <p>Forked from brave-core 5.6.3 at 3be55b5cccf881104bdd80c93e97d2575b83952d
  */
 // abstract for factory-method support on Java language level 7
```
So this is not true anymore. 😉

geminiKim replied:
Thank you for review. fiexed in 6459312.

geminiKim commented:
@trustin @minwoox Thank you for review again, I updated codes.

And I thought `samplingRate` in `LoggingDecoratorBuilder` was unnecessary and removed it. (5a8c8fd)
Does make it sense?

trustin commented on core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java:
```diff
@@ -33,6 +33,7 @@
     private static final Function<HttpHeaders, HttpHeaders> DEFAULT_HEADERS_SANITIZER = Function.identity();
     private static final Function<Object, Object> DEFAULT_CONTENT_SANITIZER = Function.identity();
     private static final Function<Throwable, Throwable> DEFAULT_CAUSE_SANITIZER = Function.identity();
+    private static final Sampler<? super RequestLog> DEFAULT_SAMPLER = Sampler.random(1.0f);
```
`Sampler.always()`

geminiKim replied:
Thank you for review. Fixed in 6d275fe.

trustin commented on core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java:
```diff
@@ -258,21 +259,18 @@ public T responseCauseSanitizer(Function<? super Throwable, ?> responseCauseSani
-     * (ex servers that each receive &lt;100K requests). If unset, all requests will be logged.
+     * Sets the {@link Sampler} that to sample requests to log.
      */
-    public T samplingRate(float samplingRate) {
```
I think we can still keep this method, because random sampling is most popular and we should avoid braking backward compatibility:

```java
public T samplingRate(float samplingRate) {
    checkArgument(...)
    return sampler(Sampler.random(samplingRate));
}
```

geminiKim replied:
Thank you for review. Updated in 307406a.

trustin commented on core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java:
```diff
@@ -258,21 +259,18 @@ public T responseCauseSanitizer(Function<? super Throwable, ?> responseCauseSani
 
     /**
-     * Returns the rate at which to sample requests to log.
+     * Returns the {@link Sampler} that to sample requests to log.
```
.. that determines which request needs logging.

geminiKim replied:
Thank you for review. Fixed in 0c5aae3.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -42,45 +42,54 @@
+@FunctionalInterface
+public interface Sampler<T> {
+    /**
+     * Returns a sampler, given a rate expressed as a percentage.
```
`expressed as a floating point number between {@code 0.0} and {@code 1.0}.`

If expressed as a percentage, it would be between 0 and 100, which is not the case.

geminiKim replied:
Thank you for review. Fixed in 0c5aae3.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -42,45 +42,54 @@
 
     /**
-     *  Returns true if a request should be recorded.
+     * Returns a sampler that never will be sampled.
```
will never be

geminiKim replied:
Thank you for review. Fixed in 0c5aae3.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -42,45 +42,54 @@
-        }
-    };
+    /**
+     * Returns a sampler that always will be sampled.
```
will always be

geminiKim replied:
Thank you for review. Fixed in 0c5aae3.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Samplers.java:
```diff
@@ -0,0 +1,55 @@
+    private Samplers() {}
+
+    /**
+     * A sampler that always will be sampled.
```
will always be

geminiKim replied:
Thank you for review. Fixed in 0c5aae3.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Samplers.java:
```diff
@@ -0,0 +1,55 @@
+    };
+
+    /**
+     * A sampler that never will be sampled.
```
will never be

geminiKim replied:
Thank you for review. Fixed in 0c5aae3.

geminiKim commented:
I have updated the code. 
Thank you for your detailed review again. 😄

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -42,45 +42,52 @@
+    /**
+     * Returns a sampler, given a rate expressed as a floating point number between {@code 0.0} and {@code 1.0}.
+     *
+     * @param rate minimum sample rate is 0.01, or 1% of traces
```
- sample rate -> sampling rate
- If the minimum sampling rate is `0.01`, then we have to say `between {@code 0.01} and {@code 1.0}`, no?

geminiKim replied:
Thank you for review. Fixed in 844219f.
(Have been fixed a similar comment in `CountingSampler`.)

trustin commented on core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java:
```diff
@@ -258,21 +261,28 @@ public T responseCauseSanitizer(Function<? super Throwable, ?> responseCauseSani
-    public T samplingRate(float samplingRate) {
-        checkArgument(0.0 <= samplingRate && samplingRate <= 1.0, "samplingRate must be between 0.0 and 1.0");
-        this.samplingRate = samplingRate;
+    public T sampler(Sampler<? super RequestContext> sampler) {
```
I guess a user cannot specify `Sampler<ClientRequestContext` or `Sampler<ServiceRequestContext>` because they are not matched by `Sampler<? super RequestContext>`. How about pushing this method down to `LoggingClientBuilder` and `LoggingServiceBuilder` so that there are two different methods:

- `LoggingClientBuilder.sampler(Sampler<? super ClientRequestContext>)`
- `LoggingServiceBuilder.sampler(Sampler<? super ServiceRequestContext>)`

You'll push down the `sampler` field and `sampler()` getter method as well.

geminiKim replied:
Thank you for review. I was mistaken that I was using 'Sampler<? extends RequestContext>'.
Fixed in 8994cfb

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -42,45 +42,52 @@
 
     /**
-     *  Returns true if a request should be recorded.
+     * Returns a sampler that will never be sampled.
```
`never be sampled` -> `always return {@code false}` ?

geminiKim replied:
Thank you for review. Fixed in 02ab62f.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java:
```diff
@@ -43,8 +43,8 @@
     private Function<Object, ?> responseContentSanitizer = DEFAULT_CONTENT_SANITIZER;
-    private Function<? super Throwable, ?> responseCauseSanitizer = DEFAULT_CAUSE_SANITIZER;
     private float samplingRate = 1.0f;
+    private Function<? super Throwable, ?> responseCauseSanitizer = DEFAULT_CAUSE_SANITIZER;
```
nit: revert?

geminiKim replied:
Thank you for review. Fixed in 02ab62f.

geminiKim commented:
@trustin @minwoox Thank you for review, but I found the problem in a failed test.

Issue point is `samplingRate` in `LoggingDecoratorBuilder`.
Now `sampler` field exist each `Logging*Builder`, But `samplingRate` exist in `LoggingDecoratorBuilder`.

In this situation, `sampler` cannot be set when using the `samplingRate (..)` method.

I think there are many ways.
- Pull-down to `LoggingServiceBuilder`. (`samplingRate` is only used by `LoggingServiceBuilder`)
- Remove `samplingRate`. (but it was already discussed in https://github.com/line/armeria/pull/1995#discussion_r320077637)
- or some...

What do you think the right ways?

geminiKim commented on core/src/main/java/com/linecorp/armeria/client/logging/LoggingClientBuilder.java:
```diff
@@ -19,15 +19,19 @@
 public class LoggingClientBuilder extends LoggingDecoratorBuilder<LoggingClientBuilder> {
+    private static final Sampler<? super ClientRequestContext> DEFAULT_SAMPLER = Sampler.always();
+
+    private Sampler<? super ClientRequestContext> sampler = DEFAULT_SAMPLER;
```
Thank you for review. Fixed in 02ab62f.

geminiKim commented on core/src/main/java/com/linecorp/armeria/server/logging/LoggingServiceBuilder.java:
```diff
@@ -21,13 +21,17 @@
 public class LoggingServiceBuilder extends LoggingDecoratorBuilder<LoggingServiceBuilder> {
+    private static final Sampler<? super ServiceRequestContext> DEFAULT_SAMPLER = Sampler.always();
+
+    private Sampler<? super ServiceRequestContext> sampler = DEFAULT_SAMPLER;
```
Thank you for review. Fixed in 02ab62f.

geminiKim commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -42,45 +42,52 @@
-        }
-    };
+    /**
+     * Returns a sampler that will always be sampled.
```
Thank you for review. Fixed in 02ab62f.

geminiKim commented on core/src/main/java/com/linecorp/armeria/common/util/CountingSampler.java:
```diff
@@ -46,20 +46,20 @@
-     * @param rate 0 means never sample, 1 means always sample. Otherwise minimum sample rate is 0.01,
-     *             or 1% of traces
+     * @param rate {@code 0.0} means never sample, {@code 1.0} means always sample. Otherwise minimum sampling
+     *               rate is between {@code 0.01} and {@code 1.0}.
```
Thank you for review. Fixed in 02ab62f.

geminiKim commented on core/src/main/java/com/linecorp/armeria/server/logging/LoggingServiceBuilder.java:
```diff
@@ -55,4 +59,19 @@
+     * Sets the {@link Sampler} that determines which request needs logging.
+     */
+    public LoggingServiceBuilder sampler(Sampler<? super ServiceRequestContext> sampler) {
+        this.sampler = sampler;
```
Thank you for review. Fixed in 02ab62f.

geminiKim commented on core/src/main/java/com/linecorp/armeria/client/logging/LoggingClientBuilder.java:
```diff
@@ -55,4 +59,19 @@
+     * Sets the {@link Sampler} that determines which request needs logging.
+     */
+    public LoggingClientBuilder sampler(Sampler<? super ClientRequestContext> sampler) {
+        this.sampler = sampler;
```
Thank you for review. Fixed in 02ab62f.

trustin commented:
@geminiKim That's a good question.

Could we pull down `samplingRate()` to `LoggingServiceBuilder` *and* `LoggingClientBuilder`? I think this method will be useful for client-side as well.

geminiKim commented:
@trustin Thank you for the advice for the right way.
Updated codes in ce59566.

trustin commented on core/src/main/java/com/linecorp/armeria/client/logging/LoggingClientBuilder.java:
```diff
@@ -55,4 +60,28 @@
+    /**
+     * Returns the {@link Sampler} that determines which request needs logging.
+     */
+    protected Sampler<? super ClientRequestContext> sampler() {
```
Maybe we do not need this method anymore, because `LoggingClientBuilder` can access it directly.

trustin commented on core/src/main/java/com/linecorp/armeria/client/logging/LoggingClientBuilder.java:
```diff
@@ -55,4 +60,28 @@
+     * Sets the rate at which to sample requests to log. Any number between {@code 0.0} and {@code 1.0} will
+     * cause a random sample of the requests to be logged.
+     */
+    public LoggingClientBuilder samplingRate(float samplingRate) {
```
Could you move the new methods (`sampler()` and `samplingRate()` before `build()` and `newDecorator()`?

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -42,45 +42,52 @@
+    /**
+     * Returns a sampler, given a rate expressed as a floating point number between {@code 0.0} and {@code 1.0}.
+     *
+     * @param rate minimum sampling rate is between {@code 0.01} and {@code 1.0}.
```
minimum sampling rate is between -> the sampling rate between

trustin commented on core/src/main/java/com/linecorp/armeria/server/logging/LoggingServiceBuilder.java:
```diff
@@ -55,4 +60,28 @@
+    /**
+     * Returns the {@link Sampler} that determines which request needs logging.
+     */
+    protected Sampler<? super ServiceRequestContext> sampler() {
```
Ditto - could be removed

trustin commented on core/src/main/java/com/linecorp/armeria/server/logging/LoggingServiceBuilder.java:
```diff
@@ -55,4 +60,28 @@
+     * Sets the rate at which to sample requests to log. Any number between {@code 0.0} and {@code 1.0} will
+     * cause a random sample of the requests to be logged.
+     */
+    public LoggingServiceBuilder samplingRate(float samplingRate) {
```
Ditto - `sampler()` and `samplingRate()` should be moved before `build() and `newDecorator()`.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -47,7 +47,7 @@
      * Returns a sampler, given a rate expressed as a floating point number between {@code 0.0} and {@code 1.0}.
      *
-     * @param rate minimum sampling rate is between {@code 0.01} and {@code 1.0}.
+     * @param rate minimum sampling rate between {@code 0.01} and {@code 1.0}.
```
minimum -> the

trustin commented on core/src/test/java/com/linecorp/armeria/common/util/RateLimitingSamplerTest.java:
```diff
@@ -0,0 +1,39 @@
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
```
Could you use JUnit 5?

trustin commented on core/src/test/java/com/linecorp/armeria/common/util/CountingSamplerTest.java:
```diff
@@ -0,0 +1,50 @@
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
```
Could you use JUnit 5?

gemini-kim replied:
Thank you for review. Fixed in 43557a8.

trustin commented on core/src/test/java/com/linecorp/armeria/common/util/CountingSamplerTest.java:
```diff
@@ -0,0 +1,50 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
```
If you forked from Brave, you must retain the copyright header of Brave here.

trustin commented on core/src/test/java/com/linecorp/armeria/common/util/RateLimitingSamplerTest.java:
```diff
@@ -0,0 +1,39 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
```
If you forked from Brave, you must retain the copyright header of Brave here.

gemini-kim replied:
Thank you for review. Fixed in 43557a8.

gemini-kim replied:
Thank you for review. Fixed in 43557a8.

gemini-kim commented:
- I've forked basic tests for `Sampler`.
- I didn't fork test-cases that used `mockStatic(PowerMock)` (https://github.com/openzipkin/brave/blob/master/brave/src/test/java/brave/sampler/RateLimitingSamplerTest.java)
- Patch test coverage is reduced, I will check it.

geminiKim commented:
@trustin  I learned that fork is not an easy task. 😭 
I am sorry and thank you for your hard work, because of my immature work and English.
Thank you again. 🙏 

(But it's not done yet, so I'll keep checking.)

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/util/CountingSampler.java:
```diff
@@ -46,20 +46,20 @@
-            return ALWAYS_SAMPLE;
+            return Sampler.always();
         }
         if (rate < 0.01f || rate > 1) {
```
nit: replace with `0.01` or `0.01D`?

geminiKim replied:
Thank you for review. Fixed in a5fe184.

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/util/RateLimitingSampler.java:
```diff
@@ -0,0 +1,184 @@
+ */
+final class RateLimitingSampler implements Sampler {
+    static Sampler create(int tracesPerSecond) {
+        if (tracesPerSecond < 0) {
```
nit: How about using `checkArgument(...)?

geminiKim replied:
Thank you for review. Fixed in a5fe184.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -42,45 +42,52 @@
  *
  * <p>Forked from brave-core 5.6.3 at 3be55b5cccf881104bdd80c93e97d2575b83952d
```
I think we should remove this line.

geminiKim replied:
Thank you for review. Fixed in d42052d.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -42,45 +42,52 @@
+    /**
+     * Returns a sampler, given a rate expressed as a floating point number between {@code 0.0} and {@code 1.0}.
+     *
+     * @param rate the sampling rate between {@code 0.01} and {@code 1.0}.
```
nit: `{@code 0.01}` -> `{@code 0.0}`?

geminiKim replied:
Thank you for review. 
I think `{@code 0.01}` is correct in this comment.
Because in the constructor of `CountingSampler` validates with a value of `0.01`.
https://github.com/line/armeria/blob/d42052d7b0d26de4fd23710cb371f73c6d23df60/core/src/main/java/com/linecorp/armeria/common/util/CountingSampler.java#L64

Please let me know if I'm wrong. Thank you! 🙏

minwoox replied:
The above doc says `number between {@code 0.0} and {@code 1.0}`, but it's weird that the parameter description says `between {@code 0.01} and {@code 1.0}`. I think we need to use the same value.

And now I figured why. `0` is a valid value but the value less than `0.01` is not valid.
```java
if (rate == 0) { // 0 is valid.
    return NEVER_SAMPLE;
}
if (rate == 1.0) {
    return ALWAYS_SAMPLE;
}
if (rate < 0.01 || rate > 1) { // What da heck?
    throw new IllegalArgumentException("rate should be between 0.01 and 1: was " + rate);
}
```
So I think we need to fix this.
```java
final class CountingSampler extends Sampler {
    public static Sampler create(final float rate) {
        final int percent = (int) (rate * 100.0f);
        if (percent == 0) {
            return NEVER_SAMPLE;
        }
        if (percent == 100) {
            return ALWAYS_SAMPLE;
        }
        if (percent < 0 || percent > 100) {
            throw new IllegalArgumentException("rate should be between 0.0 and 1: was " + rate);
        }
        return new CountingSampler(rate);
    }

	...
	CountingSampler(int percent, Random random) {
	...
```
What do you think? @geminiKim

geminiKim replied:
I think that is good idea, Thank you for review!

geminiKim commented on core/src/main/java/com/linecorp/armeria/common/util/Sampler.java:
```diff
@@ -42,45 +42,52 @@
-     * If you have high volumes of traffic, consider {@code BoundarySampler}.
-     *
-     * @param rate minimum sample rate is 0.01, or 1% of traces
+     * @param object The object to be decided on, can be ignored
```
Thank you for review. Fixed in a5fe184.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/util/CountingSampler.java:
```diff
@@ -46,23 +48,22 @@
+        if (percent == 100) {
+            return Sampler.always();
         }
         return new CountingSampler(rate);
```
I think we can change the constructor to take an integer value which is `percent` because the rate is changed to percent after all. :)

adriancole replied:
so less than a percent isn't useful? mind you that in zipkin less than a percent is common

adriancole replied:
nevermind me :) I recall we used different sampler when the percentage was below 1%

minwoox replied:
Yes, the previous logic was ignoring that anyway. :) Which sampler Zipkin uses when the percentage was below 1%? I think we should follow that logic. 🤔

adriancole replied:
The other sampler won't work for you because it is based on a pattern on random ID (Boundary Sampler). High rate can be addressed by switching to rate-based, or you can go with counting with another layer of bucket granularity. The lowest sample rate supported was .01pct and I think some were doing .03% Not sure the translation for logging, but that's the context in tracing.

minwoox replied:
Then let's keep the sampler as it is and make another one for the rate under 0.01 when there's a demand for that.
Thanks for the explanation. 😄

geminiKim replied:
@minwoox Thank you for review. I missed this because I didn't see deep to the constructor.
Fixed in cd47b3b

@adriancole Thank you for the additional information and explanation.

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/CountingSampler.java:
```diff
@@ -46,47 +48,45 @@
-            return ALWAYS_SAMPLE;
+    static Sampler create(final double rate) {
+        final int percent = (int) (rate * 100.0);
+        checkArgument(percent >= 0 && percent <= 100, "rate should be between 0.0 and 1: was " + rate);
```
1 -> 1.0 ? :sweat_smile:

trustin commented:
I added a few commits that makes some minor changes. Will merge once the build passes.

geminiKim commented:
@trustin Thank you for your work. I'm sorry for my shortage. 😭
